Turn tuples into structs

Dscano opened the pull request:
This PR attempts to address [issue #206](https://github.com/p4lang/p4mlir-incubator/issues/206)

As a beginner with MLIR, I'm still getting familiar with it . I'd really appreciate any feedback or suggestions you might have!

asl commented on lib/Transforms/TupleToStruct.cpp:
```diff
@@ -0,0 +1,123 @@
+                P4::P4MLIR::P4HIR::FieldInfo field = P4::P4MLIR::P4HIR::FieldInfo(name, elementTypes[i]);
+                structFiedls.push_back(field);
+            }
+            llvm::StringRef name("_tupletoStruct");
```
Why do you need this extra variable?

asl commented on lib/Transforms/TupleToStruct.cpp:
```diff
@@ -0,0 +1,123 @@
+      return rewriter.notifyMatchFailure(op, "Type conversion failed to struct");
+
+    Value newStruct = rewriter.create<P4::P4MLIR::P4HIR::StructOp>(loc, structType, adaptor.getOperands());
+    rewriter.replaceOp(op, newStruct);
```
Use `replaceOpWithNewOp` that constructs in-place

asl commented on lib/Transforms/TupleToStruct.cpp:
```diff
@@ -0,0 +1,123 @@
+   }
+};
+
+struct TupleExtractOpConversion : public OpConversionPattern<P4::P4MLIR::P4HIR::TupleExtractOp> {
```
Why these repeating `P4::P4MLIR`? You're already `using` this namespace above. Same in other places

asl commented on lib/Transforms/TupleToStruct.cpp:
```diff
@@ -0,0 +1,123 @@
+                                ConversionPatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+    Type converted = getTypeConverter()->convertType(op.getType());
+    auto structType = dyn_cast<P4::P4MLIR::P4HIR::StructType>(converted);
```
I'd just fold `dyn_cast` into call above. There is no use for `converted` below

asl commented on lib/Transforms/TupleToStruct.cpp:
```diff
@@ -0,0 +1,123 @@
+    auto structType = dyn_cast<P4::P4MLIR::P4HIR::StructType>(input.getType());
+
+    if(!structType)
+      return rewriter.notifyMatchFailure(op, "Expect StructType as Input");
```
Could probably proof-edit messages here and everywhere else.

asl commented on lib/Transforms/TupleToStruct.cpp:
```diff
@@ -0,0 +1,123 @@
+      return rewriter.notifyMatchFailure(op, "Expect StructType as Input");
+
+    if(index >= structType.getFields().size())
+      return rewriter.notifyMatchFailure(op, "Indexout of bounds in StructType");
```
This must be hard error actually.

asl commented on lib/Transforms/TupleToStruct.cpp:
```diff
@@ -0,0 +1,123 @@
+      return rewriter.notifyMatchFailure(op, "Indexout of bounds in StructType");
+
+    const auto &field = structType.getFields()[index];
+    StringAttr fieldName = field.name;
```
why extra variable?

asl commented on test/Transforms/Passes/tuple-to-struct.mlir:
```diff
@@ -0,0 +1,84 @@
+// CHECK-LABEL: module
+// CHECK-LABEL: p4hir.func @bid
+// CHECK: p4hir.return
+// CHECK-NOT: p4hir.tuple
```
You'd also need to match for the new struct types

asl commented on test/Transforms/Passes/tuple-to-struct.mlir:
```diff
@@ -0,0 +1,84 @@
+// CHECK-LABEL: p4hir.func @bid
+// CHECK: p4hir.return
+// CHECK-NOT: p4hir.tuple
+// CHECK-NOT: p4hir.tuple_extract
```
Looks like a copy-paste? There is neither tuple op / tuple_extract ops below. Please ensure that checks are correct.

asl commented on test/Transforms/Passes/tuple-to-struct.mlir:
```diff
@@ -0,0 +1,84 @@
+
+// CHECK-LABEL: module
+// CHECK-LABEL: p4hir.func @nested_extract
+// CHECK : p4hir.struct_extract
```
Must be two struct extracts below, no? Please ensure you're matching operands as well

asl commented on include/p4mlir/Transforms/Passes.td:
```diff
@@ -141,4 +141,19 @@ def RemoveAliases : Pass<"p4hir-remove-aliases", "mlir::ModuleOp"> {
+def TupleToStruct : Pass<"p4hir-tuple-to-struct", "mlir::ModuleOp"> {
+  let summary = "Convert p4hir.tuple types/ops into p4hir.stract types/ops";
+  let description = [{
+    Eliminate all p4hir.tuple types and operations by converting them into equivalent
```
Well, there are no `p4hir.tuple` types actually... Also, will you please correct the typos?

asl reviewed (changes requested):
See comments below. Also, will you please run `clang-format` and check for typos and spelling errors?

Thanks!

Dscano commented on lib/Transforms/TupleToStruct.cpp:
```diff
@@ -0,0 +1,123 @@
+            SmallVector<P4::P4MLIR::P4HIR::FieldInfo> structFiedls;
+            structFiedls.reserve(elementTypes.size());
+
+            for(unsigned i=0; i< elementTypes.size(); i++){
```
done

Dscano commented on lib/Transforms/TupleToStruct.cpp:
```diff
@@ -0,0 +1,123 @@
+            structFiedls.reserve(elementTypes.size());
+
+            for(unsigned i=0; i< elementTypes.size(); i++){
+                mlir::StringAttr name = mlir::StringAttr::get(ctx, "elemet_" + std::to_string(i));
```
done

Dscano commented:
@asl I apologize for the grammatical errors and for skipping `clang-format`. I’ve addressed most of your comments; next I’ll refine the tests to validate the tuple-to-struct conversion.
I also found that `TupleToStructPass` required additional code to translate call operands, apply the type converter, and update function-like operation signatures and types, similar to` SerEnumElimination`.

asl commented:
> I also found that `TupleToStructPass` required additional code to translate call operands, apply the type converter, and update function-like operation signatures and types, similar to` SerEnumElimination`.

Yeah. I'd guess we'd need some common boilerplate for this as this is a repeating pattern for type conversion

asl commented on lib/Transforms/TupleToStruct.cpp:
```diff
@@ -0,0 +1,139 @@
+        return success();
+    }
+};
+struct ReturnOpConversion : public OpConversionPattern<P4HIR::ReturnOp> {
```
why do you need this?

asl commented on lib/Transforms/TupleToStruct.cpp:
```diff
@@ -0,0 +1,131 @@
+            return rewriter.notifyMatchFailure(op, "expected P4HIR StructType as input operand");
+
+        if (index >= structType.getFields().size())
+            return rewriter.notifyMatchFailure(op, "Index out of bounds in P4HIR StructType");
```
would be great to unify the style of messages. Normally we start from lowercase letter.

asl commented on test/Transforms/Passes/tuple-to-struct.mlir:
```diff
@@ -0,0 +1,99 @@
+// RUN: p4mlir-opt --p4hir-tuple-to-struct %s | FileCheck %s
```
it is better to use `--split-input-file` here to ensure that things are really independent

asl commented on test/Transforms/Passes/tuple-to-struct.mlir:
```diff
@@ -0,0 +1,99 @@
+
+// CHECK-LABEL: module
+// CHECK-LABEL: p4hir.func @bid
+// CHECK-SAME: (%arg0: !_tupleToStruct) -> !_tupleToStruct
```
You really need to match the resulting type as type name alias in general is meaningless. So you'd do:

`![[tupleToStructType:.*]] = !p4hir.struct<"_tupleToStruct", element_0: !b32i, element_1: !i32i>`

And then match using `[[tupleToStructType]]` variable name.

asl commented on test/Transforms/Passes/tuple-to-struct.mlir:
```diff
@@ -0,0 +1,99 @@
+// CHECK-LABEL: module
+// CHECK-LABEL: p4hir.func @make_tuple
+// CHECK-SAME: (%arg0: !b32i, %arg1: !i32i) -> !_tupleToStruct
+// CHECK: %[[S:[a-zA-Z0-9_]+]] = p4hir.struct (%arg0, %arg1) : !_tupleToStruct
```
I'd not bother about `[a-zA-Z0-9_]+` pattern and use more generic `.*` :)

asl reviewed (commented):
See comments

asl commented on lib/Transforms/TupleToStruct.cpp:
```diff
@@ -99,27 +99,14 @@ void TupleToStructPass::runOnOperation() {
-        return typeConverter.isLegal(op->getOperandTypes()) &&
-               typeConverter.isLegal(op->getResultTypes());
-    });
+    patterns.add<TupleOpConversion, TupleExtractOpConversion>(typeConverter, &context);
```
While now `configureUnknownOpDynamicallyLegalByTypes` handles the things via result type, I'd just explicitly do
```c++
    addIllegalOp<P4HIR::TupleOp, P4HIR::TupleExtractOp>();
```
just to clearly outline what is expected output :)
